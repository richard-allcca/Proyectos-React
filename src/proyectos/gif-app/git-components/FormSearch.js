import { PropTypes } from "prop-types";
import React, { useState } from "react";

const FormSearch = ({ handleAdd }) => {
  const [inputValue, setInputValue] = useState("");

  const handleChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!inputValue) return;
    handleAdd(inputValue);
    setInputValue("");
  };

  return (
    <form onSubmit={handleSubmit} className="form-gif">
      <input type="text" value={inputValue} onChange={handleChange} />
      {/* <button type="submit" value="Agregar">
        Agregar
      </button> */}
    </form>
  );
};

FormSearch.prototype = {
  handleAdd: PropTypes.func.isRequired,
};

export default FormSearch;

// Notas:
// Categories.prototype: obliga recibir la funcion (handleAdd) del padre(GifApp)
