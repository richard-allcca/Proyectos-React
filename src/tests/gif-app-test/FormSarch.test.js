import { shallow } from "enzyme";
import React from "react";
import FormSearch from "./../../proyectos/gif-app/gif-components/FormSearch";

describe("Prueba de Envio de formulario", () => {
  //* funcion para test normal
  // const handleAdd = (e) => {};
  const handleAdd = jest.fn();

  let wrapper = shallow(<FormSearch handleAdd={handleAdd} />);

  beforeEach(() => {
    wrapper = shallow(<FormSearch handleAdd={handleAdd} />);
    jest.clearAllMocks();
  });

  test("Debe mostrarse correctamente el componente FormSearch", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe mostrarse correctamente valor del input", () => {
    const input = wrapper.find("input");
    const value = "Hola test mundo";

    input.simulate("change", { target: { value } });

    //? innecesario crear "p" para verificar el contenido del estado "inputValue"
    // expect(wrapper.find("p").text()).toBe(value);
  });

  test("NO Debe mostrar la info en submit del formulario", () => {
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    expect(handleAdd).not.toHaveBeenCalled();
  });

  test("DEBE llamar el handleAdd y limpiar la caja de texto", () => {
    const value = "Hola Mundo";

    // 1. simular el inputChange
    wrapper.find("input").simulate("change", { target: { value } });

    // 2. simular el submit
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    // 3. handleAdd se debe de haber llamado
    expect(handleAdd).toHaveBeenCalled();
    expect(handleAdd).toHaveBeenCalledTimes(1);
    //? expect(handleAdd).toHaveBeenCalledWith(expect.any(Function));

    // 4. el valor del input debe de estar ''
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});

// Notas;
// wrapper: tiene doble inicializacion fuera y dentro de beforeEach por ayuda de autocompletado
// jest.fn(): funcion que simula el submit de formulario
// beforeEach: reinicia el wrapper en caso le hagas cambios al component original
// en la simulaci√≥n del formulario use un 'beforeEach' normal ln/14
// el profe uso jest.clearAllMocks(), (pero yo no necesite usarlo y paso el test)
// toHaveBeenCalled asegura que la funcion ha sido llamada (.not) niega esa validacion
// toHaveBeenCalled: verifica la llamada de la fucion
// toHaveBeenCalledTimes: verifica la llamada de la fucion (X) numero de veces
// toHaveBeenCalledWith: verifica la llamada de la fucion con un argumento funcion(no en mi caso)
