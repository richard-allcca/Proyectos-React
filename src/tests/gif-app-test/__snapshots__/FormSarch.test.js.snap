// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de Envio de formulario Debe mostrarse correctamente el componente FormSearch 1`] = `
<form
  className="form-gif"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
